src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -0,0 +1,287 @@
+        } else {
+            return Hash();
+        }
+    }
+};
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::wfcheck() const {
+    if (parents.size() >= Depth) {
+        throw std::ios_base::failure("tree has too many parents");
+    }
+}
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::append(Hash obj) {
+    if (is_complete(Depth)) {
+        throw std::runtime_error("tree is full");
+    }
+
+    if (!left) {
+        // Set the left leaf
+        left = obj;
+    } else if (!right) {
+        // Set the right leaf
+        right = obj;
+    } else {
+        // Combine the leaves and propagate it up the tree
+        boost::optional<Hash> combined = Hash::combine(*left, *right);
+
+        // Set the left leaf to the object and make the right object none
+        left = obj;
+        right = boost::none;
+
+        // Propagate up the tree as much as needed
+        BOOST_FOREACH(boost::optional<Hash>& parent, parents) {
+            if (parent) {
+                combined = Hash::combine(*parent, *combined);
+                parent = boost::none;
+            } else {
+                parent = *combined;
+                combined = boost::none;
+                break;
+            }
+        }
+
+        if (combined) {
+            // Create a new parent
+            parents.push_back(combined);
+        }
+    }
+}
+
+// This is for allowing the witness to determine if a subtree has filled
+// to a particular depth, or for append() to ensure we're not appending
+// to a full tree.
+template<size_t Depth, typename Hash>
+bool IncrementalMerkleTree<Depth, Hash>::is_complete(size_t depth) const {
+    if (!left || !right) {
+        return false;
+    }
+
+    if (parents.size() != (depth - 1)) {
+        return false;
+    }
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
+        if (!parent) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+// This finds the next "depth" of an unfilled subtree, given that we've filled
+// `skip` uncles/subtrees.
+template<size_t Depth, typename Hash>
+size_t IncrementalMerkleTree<Depth, Hash>::next_depth(size_t skip) const {
+    if (!left) {
+        if (skip) { 
+            skip--;
+        } else {
+            return 0;
+        }
+    }
+
+    if (!right) {
+        if (skip) { 
+            skip--;
+        } else {
+            return 0;
+        }
+    }
+
+    size_t d = 1;
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
+        if (!parent) {
+            if (skip) {
+                skip--;
+            } else {
+                return d;
+            }
+        }
+
+        d++;
+    }
+
+    return d + skip;
+}
+
+// This calculates the root of the tree.
+template<size_t Depth, typename Hash>
+Hash IncrementalMerkleTree<Depth, Hash>::root(size_t depth,
+                                              std::deque<Hash> filler_hashes) const {
```

Comment:
The fixed point was what allowed the root of an empty tree to be zero. Getting rid of the fixed point implies computing the actual roots of an "empty" (actually all zero leafs) tree.